python/feathub/processors/flink/table_builder/redis_utils.py

```diff
@@ -22,20 +23,117 @@
+from feathub.feature_tables.sources.redis_source import (
+    NAMESPACE_KEYWORD,
+    KEYS_KEYWORD,
+    RedisSource,
+    FEATURE_NAME_KEYWORD,
+    KEY_COLUMN_PREFIX,
+)
+from feathub.feature_views.derived_feature_view import DerivedFeatureView
 from feathub.processors.constants import EVENT_TIME_ATTRIBUTE_NAME
 from feathub.processors.flink.flink_jar_utils import find_jar_lib, add_jar_to_t_env
+from feathub.processors.flink.flink_types_utils import to_flink_schema
 from feathub.processors.flink.table_builder.flink_sql_expr_utils import (
     to_flink_sql_expr,
 )
 from feathub.processors.flink.table_builder.source_sink_utils_common import (
     get_schema_from_table,
 )
+from feathub.table.schema import Schema
 from feathub.table.table_descriptor import TableDescriptor
 
 
+def get_redis_source(feature_desc: TableDescriptor) -> Optional[RedisSource]:
+    """
+    Returns the RedisSource if the feature descriptor represents a RedisSource
+    optionally followed by a series of DerivedFeatureView, which is the valid
+    syntax before a lookup join using the RedisSource. Returns None if the
+    feature descriptor does not match this syntax.
+    """
+    if isinstance(feature_desc, RedisSource):
+        return feature_desc
+    if isinstance(feature_desc, DerivedFeatureView):
+        if isinstance(feature_desc.source, str):
+            raise FeathubException(
+                f"feature descriptor {feature_desc.name} must have been resolved."
+            )
+        return get_redis_source(feature_desc.source)
+    return None
+
+
+def append_redis_key_columns(
+    table: NativeFlinkTable,
+    key_expr: str,
+    namespace: str,
+    keys: List[str],
+    feature_names: List[str],
+) -> Tuple[NativeFlinkTable, List[str]]:
+    """
+    Appends columns containing the key values to write each feature to Redis.
+    :return: A Flink Table with the appended columns, and the name of these columns.
+    """
+    key_expr_template = key_expr.replace(NAMESPACE_KEYWORD, f'"{namespace}"').replace(
+        KEYS_KEYWORD, ", ".join(keys)
+    )
+
+    key_column_names = []
+    for feature_name in feature_names:
+        key_expr = key_expr_template.replace(FEATURE_NAME_KEYWORD, f'"{feature_name}"')
+        key_column_name = KEY_COLUMN_PREFIX + feature_name
+        table = table.add_columns(
+            native_flink_expr.call_sql(to_flink_sql_expr(key_expr)).alias(
+                key_column_name
+            ),
+        )
+        key_column_names.append(key_column_name)
+
+    return table, key_column_names
+
+
+def get_table_from_redis_source(
+    t_env: StreamTableEnvironment,
+    source: RedisSource,
+) -> NativeFlinkTable:
+    if source.keys is None:
+        raise FeathubException("Redis Tables must have keys.")
+
+    add_jar_to_t_env(t_env, *_get_redis_connector_jars())
+
+    feature_names = [x for x in source.schema.field_names if x not in source.keys]
+    key_column_names = [KEY_COLUMN_PREFIX + x for x in feature_names]
```

Comment:
How about naming it `physical_key_column_names`?

Reply:
Given that we have already had `source.schema.field_names` and `feature_field_names` according to comments above, it might be better to rename it to `physical_key_field_names`.

Reply:
Sounds good.